summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,103 @@
+      val tups = Await.result(tupsOut)
+
+      tups.foreach {
+        case (k, v) => {
+          val messageIdsTracker = streamMessageIdTracker(stream)
+          val messageIds = messageIdsTracker.remove(k)
+          (messageIds.isEmpty, stream.isEmpty) match {
+            case (true, true) => in.emit(List(k, v).asJava.asInstanceOf[JList[AnyRef]])
+            case (true, false) => in.emit(stream, List(k, v).asJava.asInstanceOf[JList[AnyRef]])
+            case (false, true) => in.emit(List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+            case (false, false) => in.emit(stream, List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+          }
+        }
+      }
+    }
+  }
+
+  private def convertToSummerInputFormat(data: CMap[K, V]): CMap[Int, CMap[K, V]] = {
+    data.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+  }
+
+  private def emitData[K, V](data: Future[Traversable[(Int, CMap[K, V])]], callerfunc: (String, JList[AnyRef], scala.Any) => JList[Integer], s: String, o: scala.Any): List[Int] = {
+    var returns = MList[Int]()
+    val data_trav = Await.result(data)
+    data_trav
+      .foreach {
+        case (k, v) => {
+          val messageIdsTracker = streamMessageIdTracker(s)
+          val messageIds = messageIdsTracker.remove(k)
+          val taskIds = callerfunc(s, List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+          if (taskIds != null) returns ++= taskIds.asInstanceOf[JList[Int]].asScala.toList
+        }
+      }
+    returns.toList
+  }
+
+  private def add(tuple: (K, V), o: scala.Any, s: String) = {
+    if (o != Nil) trackMessageId(tuple, o, s)
+    addToCache(tuple, s)
+  }
+
+  private def addToCache(tuple: (K, V), s: String) = {
+    val cache = spoutCaches.get(s)
+    cache match {
+      case Some(cac) => cac.add(tuple)
+      case None => {
+        spoutCaches(s) = summerBuilder.getSummer[K, V](implicitly[Semigroup[V]])
+        spoutCaches(s).add(tuple)
+      }
+    }
+  }
+
+  private def trackMessageId(tuple: (K, V), o: scala.Any, s: String): Unit = {
+    val messageIdTracker = streamMessageIdTracker.getOrElse(s, MMap[Int, MList[Object]]())
+    var messageIds = messageIdTracker.getOrElse(summerShards.summerIdFor(tuple._1), MList())
+    messageIdTracker(summerShards.summerIdFor(tuple._1)) = (messageIds.+=(o.asInstanceOf[Object]))
+    streamMessageIdTracker(s) = messageIdTracker
+  }
+
+  def emitHelper(s: String, list: JList[AnyRef], o: scala.Any)(callerFunc: (String, JList[AnyRef], scala.Any) => JList[Integer]): JList[Integer] = {
+
+    val first: K = func(list).get(0).asInstanceOf[K]
+    val second: V = func(list).get(1).asInstanceOf[V]
+
+    val emitReturn = emitData(add((first, second), o, s).map(convertToSummerInputFormat(_)), callerFunc, s, o)
+    emitReturn.asJava.asInstanceOf[JList[Integer]]
+  }
+
+  override def emit(s: String, list: JList[AnyRef], o: scala.AnyRef): JList[Integer] = emitHelper(s, list, o)((s, l, o) => in.emit(s, l, o))
```

Comment:
These can no longer have their own callbacks.  They all need to be handled the same way, since the triggered emit can mix tuples with different streams and with/without message IDs.

The logic is basically what you already have up around line 32.